Content.Shared/RCD/Systems/RCDSystem.cs

```diff
@@ -35,312 +41,588 @@ public sealed class RCDSystem : EntitySystem
+        if (component.CachedPrototype.ConstructionRules.Contains(RcdConstructionRule.MustBuildOnSubfloor) && !mapGridData.Tile.Tile.GetContentTileDefinition().IsSubFloor)
         {
-            //Floor mode just needs the tile to be a space tile (subFloor)
-            case RcdMode.Floors:
-                if (!tile.Tile.IsEmpty)
+            if (popMsgs)
+                _popup.PopupClient(Loc.GetString("rcd-component-cannot-build-as-subfloor-not-exposed-message"), uid, user);
+
+            return false;
+        }
+
+        // Tile specific rules
+        if (component.CachedPrototype.Mode == RcdMode.ConstructTile)
+        {
+            // Check rule: Tile placement is valid
+            if (!_floors.CanPlaceTile(mapGridData.GridUid, mapGridData.Component, out var reason))
+            {
+                if (popMsgs)
+                    _popup.PopupClient(reason, uid, user);
+
+                return false;
+            }
+
+            // Check rule: Tiles can't be identical
+            if (mapGridData.Tile.Tile.GetContentTileDefinition().ID == component.CachedPrototype.Prototype)
+            {
+                if (popMsgs)
+                    _popup.PopupClient(Loc.GetString("rcd-component-cannot-build-as-tiles-are-identical"), uid, user);
+
+                return false;
+            }
+
+            // Ensure that all construction rules shared between tiles and object are checked before exiting here
+            return true;
+        }
+
+        // Object specific rules
+
+        // Check rule: The tile is unoccupied
+        var isWindow = component.CachedPrototype.ConstructionRules.Contains(RcdConstructionRule.IsWindow);
+        var constructionAngle = (component.CachedPrototype.Rotation == RcdRotation.User) ? component.ConstructionDirection.ToAngle() : Direction.South.ToAngle();
+
+        foreach (var ent in _lookup.GetLocalEntitiesIntersecting(mapGridData.Tile, -0.05f, LookupFlags.Uncontained))
+        {
+            if (isWindow && HasComp<SharedCanBuildWindowOnTopComponent>(ent))
+                continue;
+
+            if (component.CachedPrototype.CollisionMask != CollisionGroup.None && TryComp<FixturesComponent>(ent, out var fixtures))
+            {
+                foreach ((var _, var fixture) in fixtures.Fixtures)
                 {
-                    _popup.PopupClient(Loc.GetString("rcd-component-cannot-build-floor-tile-not-empty-message"), uid, user);
+                    // Continue if no collision is possible
+                    if ((fixture.CollisionLayer <= 0 || (fixture.CollisionLayer & (int) component.CachedPrototype.CollisionMask) == 0))
+                        continue;
+
+                    // Continue if our custom collision bounds are not intersected
+                    if (component.CachedPrototype.CollisionBounds != null &&
+                        !DoesCustomBoundsIntersectWithFixture(component.CachedPrototype.CollisionBounds.Value, constructionAngle, ent, fixture))
```

Comment:
I think this is a common enough occurrence we need something added to entitylookupsystem but it's probably fine for now and the pr is already large.